pedsnetdcc/id_mapping_transform.py

```diff
@@ -0,0 +1,239 @@
+import logging
+import sqlalchemy
+import time
+
+from pedsnetdcc import VOCAB_TABLES
+from pedsnetdcc.abstract_transform import Transform
+from pedsnetdcc.db import Statement, StatementList
+from pedsnetdcc.dict_logging import secs_since
+from pedsnetdcc.utils import check_stmt_err, check_stmt_data
+
+# Pre-transform ID map creation statements.
+
+new_id_count_sql = """SELECT COUNT(*)
+FROM {table_name} LEFT JOIN {map_table_name} ON {pkey_name} = site_id
+WHERE site_id IS NULL"""
+new_id_count_msg = "counting new IDs needed for {table_name}"
+
+lock_last_id_sql = """LOCK {last_id_table_name}"""
+lock_last_id_msg = "locking {table_name} last ID tracking table for update"
+
+update_last_id_sql = """UPDATE {last_id_table_name} AS new
+SET last_id = new.last_id + '{new_id_count}'::integer
+FROM {last_id_table_name} AS old RETURNING old.last_id, new.last_id"""
+update_last_id_msg = "updating {table_name} last ID tracking table to reserve new IDs"  # noqa
+
+insert_new_maps_sql = """INSERT INTO {map_table_name} (site_id, dcc_id)
+SELECT {pkey_name}, row_number() over (range unbounded preceding) + '{old_last_id}'::integer
+FROM {table_name} LEFT JOIN {map_table_name} on {pkey_name} = site_id
+WHERE site_id IS NULL"""  # noqa
+insert_new_maps_msg = "inserting new {table_name} ID mappings into map table"
+
+# Mapping and last ID table naming conventions.
+
+map_table_name_tmpl = "{table_name}_ids"
+last_id_table_name_tmpl = "dcc_{table_name}_id"
+
+logger = logging.getLogger(__name__)
+
+
+class IDMappingTransform(Transform):
+
+    @classmethod
+    def pre_transform(cls, conn_str, metadata=None, table_name=None):
+        """Generate DCC IDs for the table.
+
+        This needs to be run for every table before any tables are actually
+        transformed since they need each others new IDs to place in foreign
+        key fields.
+
+        See also Transform.pre_transform.
+        """
+
+        logger.info({'msg': 'starting ID mapping pre-transform'})
+        starttime = time.time()
+
+        # Get table object and start to build results map, which will be used
+        # throughout for formatting SQL statements.
+        table = metadata.tables[table_name]
```

Comment:
I'm removing `table_name` from the method entirely and will check for `metadata` presence.